### PR TITLE
Password reset workaround 

### PR DESCRIPTION
**Pull request by LuXeZs**
This is my first time using python so if something ain't right let me know

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -12,85 +12,163 @@
 from selenium.common.exceptions import NoSuchElementException
 from webdriver_manager.chrome import ChromeDriverManager
 from selenium.webdriver.common.by import By
+from selenium.webdriver.chrome.options import Options
 
 # Load in the credentials from .env
 load_dotenv()
 
-# Get the Chromium web driver
-driver = webdriver.Chrome(ChromeDriverManager().install())
```
Comment (Derpitron) [inline src/main.py:20]: Why was this line removed?

This line of code creates the WebDriver itself, using the ChromeDriverManager library automatically. Without it, this program will not work.

File: src/main.py [hunk 2]
```diff
@@ -12,85 +12,163 @@
...
 email = os.getenv('EMAIL')
 password = os.getenv('PASSWORD')
+token = os.getenv('TOKEN')
+newpassword = os.getenv('NEWPASSWORD')
 
-def main():
-	#Blocking cloudflare, sentry.io and discord science URLS (analytics and monitoring URLS)
+def totp(gen):
+	# Get the Chromium web driver
+	driver = webdriver.Chrome(ChromeDriverManager().install())
```
Comment (Derpitron) [inline src/main.py:28]: It is potentially inefficient to create the WebDriver inside the main function.

File: src/main.py [hunk 3]
```diff
@@ -12,85 +12,163 @@
...
+	#Blocking cloudflare and discord science URLS so they don't phone home
 	driver.execute_cdp_cmd('Network.setBlockedURLs', {
 		"urls": [
 			"a.nel.cloudflare.com/report", 
 			"https://discord.com/api/v9/science",
 			"sentry.io"
 		]
 	})
-	# Enable the network connectivity of the browser
```
Comment (Derpitron) [inline src/main.py:35]: Why were the comments removed? They help to explain the code to any readers.

**Comment by Derpitron** [comment]
This looks neat! Give me a few hours, I'll review and test the code. If all goes well, I will merge it in.

File: src/main.py (hunk 2 above)
Comment (LuXeZs) [reply src/main.py:28]: As mentioned first time do anything thing with python but I did this as to allow for the user input for type of code and if they're trying to do a password reset.

File: src/main.py [hunk 4]
```diff
@@ -165,11 +190,12 @@ def reset(gen):
 					# Close the browser.
```
Comment (LuXeZs) [inline src/main.py:190]: The password reset page doesn't require any user input such as a Captcha so I found it useful to just close the browser and reopen

File: src/main.py [hunk 5]
```diff
@@ -103,15 +110,141 @@ def main():
...
+					elapsed = time.time() - start
+					print(toColor("Invalid session ticket. The Discord login session has expired, try to run the program again.", "red"))
+					print(toColor(f"Number of tried codes this session: {totpCount}", "blue"))
+					print(toColor(f"Time elapsed for codes: {elapsed}", "blue"))
+					print(toColor(f"Number of ratelimits this session: {ratelimitCount}", "blue"))
+					# Close the browser.
+					driver.close()
+					time.sleep(1)
+					# Restart and reopen the browser.
+					reset(gen)
```
Comment (LuXeZs) [inline src/main.py:194]: The password reset page doesn't require any user input such as a Captcha so I found it useful to just close the browser and reopen.

**Comment by Trxnce** [comment]
How's this looking so far?

**Comment by Trxnce** [comment]
"[0227/140201.358:INFO:CONSOLE(152)] "[AnalyticsTrackingStore] Track: Request has been terminated
Possible causes: the network is offline, Origin is not allowed by Access-Control-Allow-Origin, the page is being unloaded, etc.", source: https://discord.com/assets/cd14ce5b3de79d42c84f.js (152)" When executing for a password reset.

**Comment by Trxnce** [comment]
Update: Now opens chrome window and gives no more errors.

**Comment by Derpitron** [comment]
Apologies for the delay, I have exams and am not able to devote time to review the PR. This should be merged in by Friday, by rights

**Comment by Derpitron** [comment]
Hello @Farid31J , this feature is not available as of yet, it will be added in around 1 week's time. Follow this conversation or Watch this repository to get updated on it.

**Comment by Farid31J** [comment]
okay, i really want to test it out but i don't know how to use it, so i guess im gonna watch the progress

File: src/main.py [hunk 6]
```diff
@@ -16,17 +16,32 @@
...
+load_dotenv(find_dotenv())
 
 # Set the email and password variables from the environment variables
 email = os.getenv('EMAIL')
 password = os.getenv('PASSWORD')
 token = os.getenv('TOKEN')
 newpassword = os.getenv('NEWPASSWORD')
 
-def totp(gen):
+def main(mode,code):
```
Comment (Derpitron) [inline src/main.py:27]: It would be better to split the individual "modes" into their own functions, e.g `def totp`, and `def pr` (Note: add comments which explain what the mode abbreviation's full forms are)

E.g:

```py
def totp(code):
	#totp logic

def pr(code):
	#password reset logic

def main():
	umode = input("Enter the program mode")
	if umode == "totp":
		totp(code)
	elif umode == "pr":
		pr(code)
```

This would make each individual feature and mode easier to maintain and debug.

File: src/main.py [hunk 7]
```diff
@@ -16,17 +16,32 @@
...
+	options.add_argument('--log-level=1')
+	options.add_experimental_option("excludeSwitches", ["enable-logging"])
+	# Sets what page to use.
+	match mode:
+		case "OTP":
+			# OTP login
+			dispage = "https://www.discord.com/login"
+		case "PR":
+			# Password reset page
+			options.add_argument('--headless')
```
Comment (Derpitron) [inline src/main.py:39]: It might be wise if the `--headless` option is removed.

This is because the user must be able to stop and monitor the process at any time, for any purpose. A headless browser hides what's going on from the user, which is not ideal especially if the user is security and ip-ban-conscious.

File: src/main.py [hunk 8]
```diff
@@ -37,54 +52,72 @@ def totp(gen):
...
-	# Go to the discord login page
-	driver.get("https://www.discord.com/login")
-	options = webdriver.ChromeOptions() 
-	# Clean up the program log
-	options.add_experimental_option("excludeSwitches", ["enable-logging"])
+	# Go to the discord page
+	driver.get(dispage)
 	# Wait 1 second before typing the email and password
 	time.sleep(1)
+	match mode:
```
Comment (Derpitron) [inline src/main.py:59]: Related: My earlier comment about refactoring the program based on functions for modes.

**Comment by Derpitron** [comment]
Why does this program need the `token` as a user input?

**Comment by LuXeZs** [comment]
> Why does this program need the `token` as a user input?

The link you get from discord to reset you password is static and looks like this https://discord.com/reset#token=
So I just ask the user what the token is so it can open the reset page for them

**Comment by LuXeZs** [comment]
> You are doing an amazing job with your code! I apologize for not being able to test it in depth and help you out. Don't get discouraged by my comments (if you are)

I'm not getting discouraged as it's my first ever time even trying to code python so learning all the little quirks has been fun and I mainly do stuff after work so I'm usually not full focused

**Comment by Derpitron** [comment]
> The link you get from discord to reset you password is static and looks like this [discord.com/reset#token=](https://discord.com/reset#token=) So I just ask the user what the token is so it can open the reset page for them

What if the user doesn't know their token?

**Comment by Farid31J** [comment]
> > The link you get from discord to reset you password is static and looks like this [discord.com/reset#token=](https://discord.com/reset#token=) So I just ask the user what the token is so it can open the reset page for them
> 
> What if the user doesn't know their token?

i think he means that the token from our reset password token but im kinda confusd too what token for it.

**Comment by LuXeZs** [comment]
> > > The link you get from discord to reset you password is static and looks like this [discord.com/reset#token=](https://discord.com/reset#token=) So I just ask the user what the token is so it can open the reset page for them
> > 
> > 
> > What if the user doesn't know their token?
> 
> i think he means that the token from our reset password token but im kinda confusd too what token for it.

It is the the link you get from the email, you have to mainly request the reset first and then copy the token and put it in the .env 

I'll add the steps to the readme when I have some free time.

**Comment by Farid31J** [comment]
> > > > The link you get from discord to reset you password is static and looks like this [discord.com/reset#token=](https://discord.com/reset#token=) So I just ask the user what the token is so it can open the reset page for them
> > > 
> > > 
> > > What if the user doesn't know their token?
> > 
> > 
> > i think he means that the token from our reset password token but im kinda confusd too what token for it.
> 
> It is the the link you get from the email, you have to mainly request the reset first and then copy the token and put it in the .env
> 
> I'll add the steps to the readme when I have some free time.

okay thanks !

**Comment by LuXeZs** [comment]
@Derpitron I'm going to work on all your suggestions today and will try explain my work better, I have been setting it up around my preference and forgetting about how easy it would be to read and understand.

Sorry everyone about the mess first time using GitHub so bear with me well I learn.

File: src/main.py (hunk 7 above)
Comment (LuXeZs) [reply src/main.py:39]: Made the option off by default with the use of a comment, so if users want to use `--headless` they just need to uncomment it.

File: src/main.py (hunk 8 above)
Comment (LuXeZs) [reply src/main.py:59]: I didn't see these before I pushed my code so that's on me sorry about that.

**Comment by Farid31J** [comment]
> Forgot to look at this page before I continued my work, so my mistake for not resolving the issues but I have got a new update with hopefully all the issues resolved.
> 
> @Derpitron I'm going to work on all your suggestions today and will try explain my work better, I have been setting it up around my preference and forgetting about how easy it would be to read and understand.
> 
> Sorry everyone about the mess first time using GitHub so bear with me well I lea

it's fine! atleast i can used it after it done, or someone can test it for me, cause im not really understand to some coding stuff

File: src/main.py [hunk 9]
```diff
@@ -103,15 +181,36 @@ def main():
...
+					print("Code: " + toColor(totp, "blue") + " did not work, delay: " + toColor(sleepy, "blue"))
 
-		# If the TOTP login field is not found (e.g the user hasn't completed the Captcha, then try again
 		except (NoSuchElementException):
 			pass
 
 if __name__ == "__main__":
-	# Run the program.
-	main()
+	try:
```
[the remark this answers is not in the document]
Comment (LuXeZs) [reply src/main.py]: This has been removed as I thought it was needed but was not.

**Comment by Farid31J** [comment]
Screenshot (78)

im trying resetting the password and try to test it

**Comment by Derpitron** [comment]
@Farid31J the program is working as intended. Though, it is not guaranteed to always get the correct code, so you might have to attempt it many times. Run it at most 4 times a day, otherwise you may get IP banned.

**Comment by Farid31J** [comment]
> @Farid31J the program is working as intended. Though, it is not guaranteed to always get the correct code, so you might have to attempt it many times. Run it at most 4 times a day, otherwise you may get IP banned.

okay!
